File: packages/source-evm/src/ws.ts
import {
  Commitment,
  CursorReorderBuffer,
  Network as NetworkValue,
  type BackfillRequest,
  type BootstrapSnapshot,
  type ChainCursor,
  type ChainDataSource,
  type ChainUpdate,
  type Checkpoint,
  type ContractFilter,
  type ContractLog,
  type DeploymentConfig,
  type Network,
} from "@lunarbase-lab/pmm-v2-client";
import { parseAddress, type Address } from "@lunarbase-lab/pmm-v2-math";
import type { Hex } from "ox/Hex";
import { JsonRpcHttpClient, parseRpcLog, RpcError, RpcHttpBackend, RpcSnapshotProvider } from "./rpc.js";
import {
  defaultWebSocketFactory,
  establishSocket,
  type BoundedFrameQueue,
  type EstablishedSocket,
  type WebSocketFactory,
} from "./ws/connection.js";
import { gap, parseHead } from "./ws/protocol.js";

const STANDARD_LOG_GRACE_MILLISECONDS = 2_000;
const STANDARD_HEAD_DEADLINE = Symbol("standard-head-deadline");

type ParsedHead = { cursor: ChainCursor; parentHash?: Hex };
type OpenStandardHead = { readonly observedAt: number; readonly head: ParsedHead };

export { BoundedFrameQueue, defaultWebSocketFactory } from "./ws/connection.js";
export type { SocketEvent, WebSocketFactory, WebSocketLike } from "./ws/connection.js";

/** Resource bounds for generic Ethereum WebSocket ingestion. */
export interface WsRpcConfig {
  /** Maximum accepted WebSocket frame size before fail-closed recovery. */
  readonly maxFrameBytes: number;
  /** Maximum total bytes retained by decoded socket frames. */
  readonly queueByteCapacity: number;
  /** Maximum total bytes retained by notifications received during handshake. */
  readonly prefetchByteCapacity: number;
  /** Maximum updates retained while awaiting an ordering watermark. */
  readonly reorderCapacity: number;
  /** Maximum total bytes retained while awaiting an ordering watermark. */
  readonly reorderByteCapacity: number;
  /** Maximum decoded frames waiting for the socket consumer. */
  readonly queueCapacity: number;
  /** Ethereum subscription method used for Core event logs. */
  readonly logsSubscription: "logs" | "pendingLogs";
  /** Whether multiple sequenced heads at one block height are valid progress. */
  readonly progressiveHeads: boolean;
}

export const DEFAULT_WS_RPC_CONFIG: WsRpcConfig = Object.freeze({
  maxFrameBytes: 256 * 1024,
  queueByteCapacity: 16 * 1024 * 1024,
  prefetchByteCapacity: 16 * 1024 * 1024,
  reorderCapacity: 4096,
  reorderByteCapacity: 64 * 1024 * 1024,
  queueCapacity: 4096,
  logsSubscription: "logs",
  progressiveHeads: false,
});

/**
 * Standard Ethereum JSON-RPC WebSocket source. HTTP remains authoritative
 * for block-tagged bootstrap and canonical backfill; a socket gap is emitted
 * instead of being silently hidden by reconnecting from an unknown cursor.
 */
export class EvmRpcSource implements ChainDataSource {
  /** Canonical HTTP backend used for heads, backfill, and validation. */
  private readonly http: RpcHttpBackend;
  /** Coherent block-tagged Core snapshot provider. */
  private readonly snapshots: RpcSnapshotProvider;
  /** Platform or injected WebSocket constructor. */
  private readonly factory: WebSocketFactory;
  /** Validated frame, queue, ordering, and subscription limits. */
  readonly config: WsRpcConfig;

  /** Creates a WebSocket backend with bounded frame, queue, and reorder memory. */
  constructor(
    /** Strict read-only HTTP client used outside the quote path. */
    readonly rpc: JsonRpcHttpClient,
    /** WebSocket endpoint used exclusively for realtime subscriptions. */
    readonly wsEndpoint: string,
    /** Network family exposed through the common source interface. */
    readonly network: Network,
    /** EIP-155 chain identifier attached to normalized cursors. */
    readonly chainId: bigint,
    /** Canonical block tag used for snapshot and recovery reads. */
    readonly snapshotTag = "latest",
    config: Partial<WsRpcConfig> = {},
    factory: WebSocketFactory = defaultWebSocketFactory,
  ) {
    this.config = validateConfig({ ...DEFAULT_WS_RPC_CONFIG, ...config });
    this.factory = factory;
    this.http = new RpcHttpBackend(rpc, network, chainId, snapshotTag);
    this.snapshots = new RpcSnapshotProvider(rpc, snapshotTag);
  }

  /** Reads one coherent quote state through block-tagged HTTP calls. */
  snapshot(deployment: DeploymentConfig): Promise<BootstrapSnapshot> {
    if (deployment.network !== this.network)
      return Promise.reject(new RpcError("INVALID", "RPC source network mismatch"));
    if (deployment.chainId !== this.chainId)
      return Promise.reject(new RpcError("INVALID", "RPC source chain id mismatch"));
    return this.snapshots.snapshot(deployment);
  }

  /** Delegates canonical log backfill to HTTP. */
  backfill(request: BackfillRequest): Promise<readonly ContractLog[]> {
    return this.http.backfill(request);
  }

  /**
   * Opens and acknowledges logs/new-head subscriptions before exposing the
   * stream. A client therefore cannot become ready while the socket is merely
   * connecting or while the node has rejected either subscription.
   */
  async subscribe(filter: ContractFilter, signal?: AbortSignal): Promise<AsyncIterable<ChainUpdate>> {
    const expectedAddress = parseAddress(filter.address);
    const connection = establishSocket(
      this.wsEndpoint,
      this.factory,
      filter,
      this.config.logsSubscription,
      this.chainId,
      this.config.queueCapacity,
      this.config.queueByteCapacity,
      this.config.prefetchByteCapacity,
      this.config.maxFrameBytes,
      signal,
    );
    try {
      const [established] = await Promise.all([connection, this.http.verifyChainId()]);
      return this.readSocket(established, expectedAddress, signal);
    } catch (error) {
      void connection.then(
        (established) => established.close(),
        () => undefined,
      );
      throw error;
    }
  }

  /** Returns the canonical HTTP recovery head. */
  canonicalHead(): Promise<ChainCursor> {
    return this.http.canonicalHead();
  }

  /** Verifies a checkpoint block hash through canonical HTTP RPC. */
  validateCheckpoint(checkpoint: Checkpoint): Promise<boolean> {
    return this.http.validateCheckpoint(checkpoint);
  }

  private async *readSocket(
    connection: EstablishedSocket,
    expectedAddress: Address,
    signal?: AbortSignal,
  ): AsyncIterable<ChainUpdate> {
    const { queue, logsSubscription, headsSubscription, prefetched, close } = connection;
    let lastHead: ChainCursor | undefined;
    let lastParentHash: Hex | undefined;
    let sourceSequence = 0n;
    let reorder = new CursorReorderBuffer(this.config.reorderCapacity, this.config.reorderByteCapacity);
    const standardLogs = holdsStandardLogsUntilSuccessor(this.config);
    let openStandardHeads: OpenStandardHead[] = [];
    let firstStandardHeadBlock: bigint | undefined;
    let publishedWatermark: ChainCursor | undefined;

    try {
      while (!signal?.aborted) {
        let frame: string | undefined | typeof STANDARD_HEAD_DEADLINE;
        try {
          frame =
            prefetched.shift() ??
            (await nextFrameOrStandardDeadline(
              queue,
              standardLogs ? standardHeadDeadline(openStandardHeads) : undefined,
              signal,
            ));
        } catch (error) {
          yield gap(`RPC WebSocket failed; canonical recovery required: ${message(error)}`, lastHead);
          return;
        }
        if (frame === STANDARD_HEAD_DEADLINE) {
          const completedHead = takeReadyStandardHead(openStandardHeads, monotonicMilliseconds());
          if (completedHead === undefined) continue;
          let updates: ChainUpdate[];
          try {
            updates = drainCompletedBlock(
              reorder,
              completedHead,
              firstStandardHeadBlock === completedHead.cursor.blockNumber,
            );
            updates = await validatePrecedingStartupLogs(
              updates,
              completedHead,
              this.rpc,
              this.chainId,
              this.network === NetworkValue.Arbitrum,
            );
          } catch (error) {
            yield gap(`RPC completed block failed; canonical recovery required: ${message(error)}`, lastHead);
            return;
          }
          for (const update of updates) yield update;
          publishedWatermark = completedHead.cursor;
          continue;
        }
        if (frame === undefined) {
          if (signal?.aborted) return;
          yield gap("RPC WebSocket closed; canonical recovery required", lastHead);
          return;
        }
        let value: Record<string, unknown>;
        try {
          value = JSON.parse(frame) as Record<string, unknown>;
        } catch (error) {
          yield gap(`invalid RPC WebSocket JSON; canonical recovery required: ${message(error)}`, lastHead);
          return;
        }
        if (value.error) {
          yield gap(`RPC subscription error: ${JSON.stringify(value.error)}`, lastHead);
          return;
        }
        if (value.method !== "eth_subscription" || !value.params || typeof value.params !== "object") continue;
        const params = value.params as Record<string, unknown>;
        const subscription = typeof params.subscription === "string" ? params.subscription : undefined;
        if (!subscription || params.result === undefined) {
          yield gap("RPC subscription notification is missing subscription/result", lastHead);
          return;
        }

        if (subscription === logsSubscription) {
          let log: ContractLog;
          try {
            log = parseRpcLog(params.result, this.chainId, Commitment.Realtime);
          } catch (error) {
            yield gap(`invalid RPC log notification: ${message(error)}`, lastHead);
            return;
          }
          if (log.address !== expectedAddress) {
            yield gap(`RPC log address mismatch: expected ${expectedAddress}, got ${log.address}`, log.cursor);
            return;
          }
          sourceSequence += 1n;
          log = { ...log, cursor: { ...log.cursor, sourceSequence } };
          if (log.removed) {
            yield gap("RPC retracted a subscription log; canonical recovery required", log.cursor);
            return;
          }
          if (standardLogs && isAtOrBeforeWatermark(log.cursor, publishedWatermark)) {
            yield gap("RPC delivered a log after its block watermark; canonical recovery required", log.cursor);
            return;
          }
          if (lastHead?.blockNumber === log.cursor.blockNumber)
            log = withExecutionContext(log, lastHead.executionBlockNumber);
          try {
            reorder.push({ kind: "Log", log });
          } catch (error) {
            yield gap(`RPC reorder buffer failed: ${message(error)}`, lastHead);
            return;
          }
          if (!standardLogs && lastHead)
            for (const update of reorder.drainThrough(lastHead)) yield annotateExecutionContext(update, lastHead);
          continue;
        }

        if (subscription !== headsSubscription) continue;
        let parsed: ParsedHead;
        try {
          parsed = parseHead(params.result, this.chainId, this.network === NetworkValue.Arbitrum);
        } catch (error) {
          yield gap(`invalid RPC head notification: ${message(error)}`, lastHead);
          return;
        }
        if (lastHead && sameHead(lastHead, lastParentHash, parsed)) continue;
        sourceSequence += 1n;
        parsed.cursor = { ...parsed.cursor, sourceSequence };
        if (standardLogs && firstStandardHeadBlock === undefined) firstStandardHeadBlock = parsed.cursor.blockNumber;
        if (lastHead && parsed.cursor.blockNumber > lastHead.blockNumber + 1n) {
          yield gap("RPC WebSocket skipped one or more block heads; canonical recovery required", lastHead);
          return;
        }
        if (lastHead && headDiscontinuity(lastHead, lastParentHash, parsed, this.config.progressiveHeads)) {
          yield { kind: "Reorg", oldHead: lastHead, newHead: parsed.cursor };
          reorder = new CursorReorderBuffer(this.config.reorderCapacity, this.config.reorderByteCapacity);
          openStandardHeads = [];
          firstStandardHeadBlock = parsed.cursor.blockNumber;
        }
        lastHead = parsed.cursor;
        lastParentHash = parsed.parentHash;

        if (standardLogs) {
          try {
            observeStandardHead(
              openStandardHeads,
              parsed,
              monotonicMilliseconds(),
              this.config.reorderCapacity,
              this.config.reorderByteCapacity,
            );
          } catch (error) {
            yield gap(`RPC pending heads failed: ${message(error)}`, lastHead);
            return;
          }
        }

        try {
          reorder.push({ kind: "Head", cursor: parsed.cursor });
        } catch (error) {
          yield gap(`RPC reorder buffer failed: ${message(error)}`, lastHead);
          return;
        }
        if (!standardLogs)
          for (const update of reorder.drainThrough(parsed.cursor))
            yield annotateExecutionContext(update, parsed.cursor);
      }
    } finally {
      close();
    }
  }
}

function validateConfig(config: WsRpcConfig): WsRpcConfig {
  for (const [name, value] of Object.entries({
    maxFrameBytes: config.maxFrameBytes,
    queueByteCapacity: config.queueByteCapacity,
    prefetchByteCapacity: config.prefetchByteCapacity,
    reorderCapacity: config.reorderCapacity,
    reorderByteCapacity: config.reorderByteCapacity,
    queueCapacity: config.queueCapacity,
  }))
    if (!Number.isSafeInteger(value) || value <= 0) throw new Error(`WS ${name} must be a positive safe integer`);
  if (config.logsSubscription !== "logs" && config.logsSubscription !== "pendingLogs")
    throw new Error("WS logsSubscription must be logs or pendingLogs");
  return Object.freeze(config);
}

function sameHead(lastHead: ChainCursor, lastParentHash: Hex | undefined, next: ParsedHead): boolean {
  return (
    next.cursor.blockNumber === lastHead.blockNumber &&
    next.cursor.executionBlockNumber === lastHead.executionBlockNumber &&
    next.cursor.blockHash === lastHead.blockHash &&
    next.parentHash === lastParentHash
  );
}

function headDiscontinuity(
  lastHead: ChainCursor,
  lastParentHash: Hex | undefined,
  next: ParsedHead,
  progressiveHeads: boolean,
): boolean {
  if (sameHead(lastHead, lastParentHash, next)) return false;
  if (next.cursor.blockNumber > lastHead.blockNumber + 1n) return true;
  const sameHeight = next.cursor.blockNumber === lastHead.blockNumber;
  const changedParent =
    next.parentHash !== undefined && lastParentHash !== undefined && next.parentHash !== lastParentHash;
  return (
    next.cursor.blockNumber < lastHead.blockNumber ||
    (sameHeight && (!progressiveHeads || changedParent)) ||
    (next.cursor.blockNumber === lastHead.blockNumber + 1n &&
      next.parentHash !== undefined &&
      lastHead.blockHash !== undefined &&
      next.parentHash !== lastHead.blockHash)
  );
}

function holdsStandardLogsUntilSuccessor(config: WsRpcConfig): boolean {
  return config.logsSubscription === "logs";
}

function isAtOrBeforeWatermark(cursor: ChainCursor, watermark: ChainCursor | undefined): boolean {
  return watermark !== undefined && cursor.blockNumber <= watermark.blockNumber;
}

function observeStandardHead(
  openHeads: OpenStandardHead[],
  head: ParsedHead,
  observedAt: number,
  countCapacity: number,
  byteCapacity: number,
): void {
  const nextCount = openHeads.length + 1;
  if (nextCount > countCapacity || nextCount * 256 > byteCapacity)
    throw new Error("pending head count or byte budget exceeded");
  openHeads.push({ observedAt, head });
}

function standardHeadDeadline(openHeads: readonly OpenStandardHead[]): number | undefined {
  const successor = openHeads[1];
  return successor === undefined ? undefined : successor.observedAt + STANDARD_LOG_GRACE_MILLISECONDS;
}

function takeReadyStandardHead(openHeads: OpenStandardHead[], observedAt: number): ParsedHead | undefined {
  const deadline = standardHeadDeadline(openHeads);
  if (deadline === undefined || observedAt < deadline) return undefined;
  return openHeads.shift()?.head;
}

async function nextFrameOrStandardDeadline(
  queue: BoundedFrameQueue,
  deadline: number | undefined,
  signal?: AbortSignal,
): Promise<string | undefined | typeof STANDARD_HEAD_DEADLINE> {
  if (deadline === undefined) return queue.next(signal);
  if (signal?.aborted) return undefined;

  const wait = new AbortController();
  const abort = () => wait.abort();
  signal?.addEventListener("abort", abort, { once: true });
  if (signal?.aborted) wait.abort();
  let deadlineElapsed = false;
  const timer = setTimeout(
    () => {
      deadlineElapsed = true;
      wait.abort();
    },
    Math.max(0, deadline - monotonicMilliseconds()),
  );
  try {
    const frame = await queue.next(wait.signal);
    return deadlineElapsed ? STANDARD_HEAD_DEADLINE : frame;
  } finally {
    clearTimeout(timer);
    signal?.removeEventListener("abort", abort);
  }
}

function drainCompletedBlock(
  reorder: CursorReorderBuffer,
  head: ParsedHead,
  allowPrecedingStartupLogs: boolean,
): ChainUpdate[] {
  const blockHash = head.cursor.blockHash;
  if (blockHash === undefined) throw new Error("completed RPC head has no block hash");

  const logs: ChainUpdate[] = [];
  let completedHead: ChainUpdate | undefined;
  for (const update of reorder.drainThrough(head.cursor)) {
    if (
      update.kind === "Head" &&
      update.cursor.blockNumber === head.cursor.blockNumber &&
      sameHash(update.cursor.blockHash, blockHash)
    ) {
      completedHead = update;
      continue;
    }
    if (
      update.kind === "Log" &&
      update.log.cursor.blockNumber === head.cursor.blockNumber &&
      sameHash(update.log.cursor.blockHash, blockHash)
    ) {
      logs.push(annotateExecutionContext(update, head.cursor));
      continue;
    }
    if (
      update.kind === "Log" &&
      allowPrecedingStartupLogs &&
      update.log.cursor.blockNumber < head.cursor.blockNumber &&
      update.log.cursor.blockHash !== undefined
    ) {
      logs.push(update);
      continue;
    }
    if (update.kind === "Gap") {
      const block = update.cursor?.blockNumber ?? 0n;
      throw new Error(`buffered gap at block ${block} cannot complete RPC block ${head.cursor.blockNumber}`);
    }
    const kind = update.kind.toLowerCase();
    const cursor = update.kind === "Log" ? update.log.cursor : update.kind === "Reorg" ? update.newHead : update.cursor;
    throw new Error(
      `buffered RPC ${kind} at block ${cursor.blockNumber} hash ${String(cursor.blockHash)} does not match completed block ${head.cursor.blockNumber} hash ${blockHash}`,
    );
  }
  if (completedHead === undefined) throw new Error("completed RPC block has no matching buffered head");
  return [...logs, completedHead];
}

async function validatePrecedingStartupLogs(
  updates: ChainUpdate[],
  completedHead: ParsedHead,
  rpc: JsonRpcHttpClient,
  chainId: bigint,
  requireExecutionBlockNumber: boolean,
): Promise<ChainUpdate[]> {
  const canonicalBlocks = new Map<bigint, ChainCursor>();
  for (let index = 0; index < updates.length; index += 1) {
    const update = updates[index];
    if (update?.kind !== "Log" || update.log.cursor.blockNumber >= completedHead.cursor.blockNumber) continue;
    const blockNumber = update.log.cursor.blockNumber;
    let canonical = canonicalBlocks.get(blockNumber);
    if (canonical === undefined) {
      canonical = await rpc.blockCursor(
        `0x${blockNumber.toString(16)}`,
        chainId,
        Commitment.Canonical,
        requireExecutionBlockNumber,
      );
      canonicalBlocks.set(blockNumber, canonical);
    }
    if (!sameHash(canonical.blockHash, update.log.cursor.blockHash))
      throw new Error(`startup RPC log at block ${blockNumber} does not match its canonical block hash`);
    updates[index] = {
      kind: "Log",
      log: withExecutionContext(update.log, canonical.executionBlockNumber),
    };
  }
  return updates;
}

function sameHash(left: Hex | undefined, right: Hex | undefined): boolean {
  return left !== undefined && right !== undefined && left.toLowerCase() === right.toLowerCase();
}

function monotonicMilliseconds(): number {
  return globalThis.performance.now();
}

function annotateExecutionContext(update: ChainUpdate, head: ChainCursor): ChainUpdate {
  if (update.kind !== "Log" || update.log.cursor.blockNumber !== head.blockNumber) return update;
  return { kind: "Log", log: withExecutionContext(update.log, head.executionBlockNumber) };
}

function withExecutionContext(log: ContractLog, executionBlockNumber: bigint): ContractLog {
  return { ...log, cursor: { ...log.cursor, executionBlockNumber } };
}

function message(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: crates/lunarbase-source-evm/src/ws.rs
//! Generic Ethereum JSON-RPC WebSocket source.
//!
//! The HTTP endpoint remains the authority for block-tagged snapshots and
//! canonical backfill. Standard EVM `logs` and `newHeads` subscriptions are
//! not ordered relative to each other, so a normal EVM profile closes block
//! `N` only after a later head arrives and a bounded grace period has elapsed.
//! Provider-specific pending-log profiles may consume subscription logs
//! immediately. A closed socket is deliberately terminal: callers must
//! recover from the HTTP source before claiming freshness again.

use crate::rpc::backend::RpcHttpBackend;
use crate::rpc::codec::parse_filtered_rpc_log;
use crate::rpc::snapshot::RpcSnapshotProvider;
use async_stream::stream;
use futures_util::StreamExt;
use lunarbase_client::{
    bootstrap::BootstrapSnapshot,
    model::{
        BackfillRequest, ChainCursor, ChainUpdate, Checkpoint, Commitment, ContractFilter,
        ContractLog, DeploymentConfig, Network, SourceError,
    },
    source::{ChainDataSource, SourceStream},
    state::ordering::CursorReorderBuffer,
};
use serde_json::Value;
use std::{collections::VecDeque, sync::Arc, time::Instant};

mod config;
mod connection;
mod ordering;
pub(crate) mod protocol;

pub use config::{EvmDeliveryMode, WsRpcConfig};

use connection::{establish, websocket_payload};
use ordering::{
    backfill_pages, drain_completed_block, is_at_or_before_watermark, observe_standard_head,
    promote_updates, retraction_updates, standard_head_deadline, take_ready_standard_head,
    validate_finalized_advance, validate_finalized_page, validate_preceding_startup_logs,
    with_execution_context,
};
use protocol::{WsHead, head_discontinuity, parse_ws_head_with_execution_context, same_head};

/// Generic EVM source with HTTP authority and WebSocket realtime updates.
#[derive(Clone)]
pub struct EvmRpcSource {
    /// Canonical HTTP backend used for snapshots, backfills, and validation.
    http: RpcHttpBackend,
    /// WebSocket endpoint used exclusively for realtime subscriptions.
    ws_endpoint: Arc<str>,
    /// Frame, ordering, and subscription behavior limits.
    config: WsRpcConfig,
}

impl ChainDataSource for EvmRpcSource {
    fn network(&self) -> Network {
        self.http.network()
    }

    async fn snapshot(
        &self,
        deployment: &DeploymentConfig,
    ) -> Result<BootstrapSnapshot, SourceError> {
        self.validate_config()?;
        if deployment.network != self.http.network() {
            return Err(SourceError::NetworkMismatch);
        }
        if deployment.chain_id != self.http.chain_id() {
            return Err(SourceError::Unavailable(
                "RPC source chain id mismatch".into(),
            ));
        }
        RpcSnapshotProvider::new(self.http.rpc().clone(), self.http.snapshot_tag().to_owned())
            .snapshot(deployment)
            .await
    }

    async fn backfill(&self, request: BackfillRequest) -> Result<Vec<ContractLog>, SourceError> {
        self.validate_config()?;
        self.http.backfill(request).await
    }

    async fn subscribe(&self, filter: ContractFilter) -> Result<SourceStream, SourceError> {
        self.validate_config()?;
        let (_, established) = tokio::try_join!(
            self.http.verify_chain_id(),
            establish(
                self.ws_endpoint.as_ref(),
                &filter,
                &self.config.logs_subscription,
                self.http.chain_id(),
                self.config.max_frame_bytes,
                self.config.reorder_capacity,
                self.config.max_prefetch_bytes,
            ),
        )?;
        let logs_subscription = established.logs_subscription;
        let heads_subscription = established.heads_subscription;
        let mut buffered = established.buffered;
        let socket = established.socket;
        let (mut writer, mut reader) = socket.split();

        let chain_id = self.http.chain_id();
        let http = self.http.clone();
        let config = self.config.clone();
        let initial_finalized = if config.delivery_mode == EvmDeliveryMode::Finalized {
            Some(http.snapshot_cursor(http.network()).await?)
        } else {
            None
        };
        let stream = stream! {
            let mut reorder = match CursorReorderBuffer::with_limits(
                config.reorder_capacity,
                config.reorder_byte_capacity,
            ) {
                Ok(buffer) => buffer,
                Err(error) => {
                    yield Err(error);
                    return;
                }
            };
            let mut last_head: Option<WsHead> = None;
            let mut open_standard_heads = VecDeque::new();
            let mut first_standard_head_block = None;
            let mut published_watermark: Option<ChainCursor> = None;
            let mut finalized_watermark = initial_finalized;
            let mut source_sequence = 0_u64;

            loop {
                let payload = if let Some(payload) = buffered.pop_front() {
                    payload
                } else {
                    let message = if config.holds_standard_logs_until_successor()
                        && let Some(deadline) = standard_head_deadline(&open_standard_heads)
                    {
                        tokio::select! {
                            biased;
                            () = tokio::time::sleep_until(deadline.into()) => {
                                let Some(completed_head) =
                                    take_ready_standard_head(&mut open_standard_heads, Instant::now())
                                else {
                                    continue;
                                };
                                let mut updates = match drain_completed_block(
                                    &mut reorder,
                                    &completed_head,
                                    first_standard_head_block == Some(completed_head.cursor.block_number),
                                ) {
                                    Ok(updates) => updates,
                                    Err(error) => {
                                        yield Err(error);
                                        break;
                                    }
                                };
                                if let Err(error) = validate_preceding_startup_logs(
                                    &mut updates,
                                    &completed_head,
                                    &http,
                                )
                                .await
                                {
                                    yield Err(error);
                                    break;
                                }
                                promote_updates(&mut updates, Commitment::Canonical);
                                for update in updates {
                                    yield Ok(update);
                                }
                                published_watermark = Some(completed_head.cursor);
                                continue;
                            },
                            message = reader.next() => message,
                        }
                    } else {
                        reader.next().await
                    };
                    let Some(message) = message else {
                        yield Ok(ChainUpdate::Gap {
                            cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                            reason: "RPC WebSocket closed; canonical recovery required".into(),
                        });
                        break;
                    };
                    let message = match message {
                        Ok(message) => message,
                        Err(error) => {
                            yield Ok(ChainUpdate::Gap {
                                cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                                reason: format!("RPC WebSocket read failed; canonical recovery required: {error}"),
                            });
                            break;
                        }
                    };
                    match websocket_payload(message, &mut writer).await {
                        Ok(Some(payload)) => payload,
                        Ok(None) => continue,
                        Err(error) => {
                            yield Ok(ChainUpdate::Gap {
                                cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                                reason: error.to_string(),
                            });
                            break;
                        }
                    }
                };
                if payload.len() > config.max_frame_bytes {
                    yield Ok(ChainUpdate::Gap {
                        cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                        reason: "RPC WebSocket frame exceeded configured bound".into(),
                    });
                    break;
                }
                let value: Value = match serde_json::from_slice(&payload) {
                    Ok(value) => value,
                    Err(error) => {
                        yield Ok(ChainUpdate::Gap {
                            cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                            reason: format!("invalid RPC WebSocket JSON; canonical recovery required: {error}"),
                        });
                        break;
                    }
                };

                if let Some(error) = value.get("error") {
                    yield Err(SourceError::Unavailable(format!("RPC subscription error: {error}")));
                    break;
                }
                if value.get("id").is_some() {
                    continue;
                }
                let Some(params) = value.get("params").and_then(Value::as_object) else {
                    continue;
                };
                if value.get("method").and_then(Value::as_str) != Some("eth_subscription") {
                    continue;
                }
                let Some(subscription) = params.get("subscription").and_then(Value::as_str) else {
                    yield Ok(ChainUpdate::Gap {
                        cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                        reason: "RPC subscription notification has no subscription id".into(),
                    });
                    break;
                };
                let Some(result) = params.get("result") else {
                    yield Ok(ChainUpdate::Gap {
                        cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                        reason: "RPC subscription notification has no result".into(),
                    });
                    break;
                };

                if logs_subscription == subscription {
                    if config.delivery_mode == EvmDeliveryMode::Finalized {
                        continue;
                    }
                    let mut log = match parse_filtered_rpc_log(
                        result,
                        chain_id,
                        config.delivery_mode.commitment(),
                        &filter,
                    ) {
                        Ok(log) => log,
                        Err(error) => {
                            yield Err(SourceError::Unavailable(format!("invalid RPC log notification: {error}")));
                            break;
                        }
                    };
                    source_sequence = source_sequence.saturating_add(1);
                    log.cursor.source_sequence = Some(source_sequence);
                    if log.removed {
                        for update in retraction_updates(log) {
                            yield Ok(update);
                        }
                        break;
                    }
                    if config.holds_standard_logs_until_successor()
                        && is_at_or_before_watermark(&log.cursor, published_watermark.as_ref())
                    {
                        yield Ok(ChainUpdate::Gap {
                            cursor: Some(log.cursor),
                            reason: "RPC delivered a log after its block watermark; canonical recovery required".into(),
                        });
                        break;
                    }
                    if let Some(head) = last_head.as_ref()
                        && head.cursor.block_number == log.cursor.block_number
                    {
                        log.cursor.execution_block_number =
                            head.cursor.execution_block_number;
                    }
                    if config.delivery_mode == EvmDeliveryMode::Realtime {
                        yield Ok(ChainUpdate::Log(log));
                        continue;
                    }
                    let update = ChainUpdate::Log(log);
                    match reorder.push(update) {
                        Ok(_) => {}
                        Err(error) => {
                            yield Err(error);
                            break;
                        }
                    }
                    if !config.holds_standard_logs_until_successor()
                        && let Some(head) = last_head.as_ref()
                    {
                        for update in reorder.drain_through(&head.cursor) {
                            yield Ok(with_execution_context(update, &head.cursor));
                        }
                    }
                    continue;
                }
                if heads_subscription == subscription {
                    let mut head = match parse_ws_head_with_execution_context(
                        result,
                        chain_id,
                        http.network() == Network::Arbitrum,
                    ) {
                        Ok(head) => head,
                        Err(error) => {
                            yield Err(SourceError::Unavailable(format!("invalid RPC head notification: {error}")));
                            break;
                        }
                    };
                    if last_head
                        .as_ref()
                        .is_some_and(|previous| same_head(previous, &head))
                    {
                        continue;
                    }
                    if config.delivery_mode == EvmDeliveryMode::Finalized {
                        last_head = Some(head);
                        let finalized = match http.snapshot_cursor(http.network()).await {
                            Ok(cursor) => cursor,
                            Err(error) => {
                                yield Err(error);
                                break;
                            }
                        };
                        let previous = finalized_watermark
                            .as_ref()
                            .expect("finalized mode initializes its watermark");
                        match validate_finalized_advance(previous, &finalized) {
                            Ok(false) => {
                                yield Ok(ChainUpdate::Head(finalized));
                                continue;
                            }
                            Err(error) => {
                                yield Err(error);
                                break;
                            }
                            Ok(true) => {}
                        }
                        let from_block = previous.block_number.saturating_add(1);
                        for page in backfill_pages(
                            from_block,
                            finalized.block_number,
                            config.backfill_page_blocks,
                        ) {
                            let request = BackfillRequest {
                                from_block: *page.start(),
                                to_block: *page.end(),
                                filter: filter.clone(),
                            };
                            let logs = match http.backfill(request).await {
                                Ok(logs) => logs,
                                Err(error) => {
                                    yield Err(error);
                                    return;
                                }
                            };
                            let logs = match validate_finalized_page(logs, &page) {
                                Ok(logs) => logs,
                                Err(error) => {
                                    yield Err(error);
                                    return;
                                }
                            };
                            for log in logs {
                                yield Ok(ChainUpdate::Log(log));
                            }
                        }
                        yield Ok(ChainUpdate::Head(finalized.clone()));
                        finalized_watermark = Some(finalized);
                        continue;
                    }
                    source_sequence = source_sequence.saturating_add(1);
                    head.cursor.source_sequence = Some(source_sequence);
                    head.cursor.commitment = config.delivery_mode.commitment();
                    if config.holds_standard_logs_until_successor()
                        && first_standard_head_block.is_none()
                    {
                        first_standard_head_block = Some(head.cursor.block_number);
                    }
                    if let Some(previous) = last_head.as_ref() {
                        if head.cursor.block_number
                            > previous.cursor.block_number.saturating_add(1)
                        {
                            yield Ok(ChainUpdate::Gap {
                                cursor: Some(previous.cursor.clone()),
                                reason: "RPC WebSocket skipped one or more block heads; canonical recovery required".into(),
                            });
                            break;
                        }
                        if head_discontinuity(previous, &head, config.progressive_heads) {
                            yield Ok(ChainUpdate::Reorg {
                                old_head: previous.cursor.clone(),
                                new_head: head.cursor.clone(),
                            });
                            reorder = match CursorReorderBuffer::with_limits(
                                config.reorder_capacity,
                                config.reorder_byte_capacity,
                            ) {
                                Ok(buffer) => buffer,
                                Err(error) => {
                                    yield Err(error);
                                    break;
                                }
                            };
                            open_standard_heads.clear();
                            first_standard_head_block = Some(head.cursor.block_number);
                        }
                    }
                    last_head = Some(head.clone());
                    if config.delivery_mode == EvmDeliveryMode::Realtime {
                        yield Ok(ChainUpdate::Head(head.cursor));
                        continue;
                    }
                    if config.holds_standard_logs_until_successor()
                        && let Err(error) = observe_standard_head(
                            &mut open_standard_heads,
                            head.clone(),
                            Instant::now(),
                            config.reorder_capacity,
                            config.reorder_byte_capacity,
                        )
                    {
                        yield Err(error);
                        break;
                    }
                    if let Err(error) = reorder.push(ChainUpdate::Head(head.cursor.clone())) {
                        yield Err(error);
                        break;
                    }
                    if !config.holds_standard_logs_until_successor() {
                        for update in reorder.drain_through(&head.cursor) {
                            yield Ok(with_execution_context(update, &head.cursor));
                        }
                    }
                }
            }
        };
        Ok(Box::pin(stream))
    }

    async fn canonical_head(&self) -> Result<ChainCursor, SourceError> {
        self.validate_config()?;
        self.http.snapshot_cursor(self.http.network()).await
    }

    async fn validate_checkpoint(&self, checkpoint: &Checkpoint) -> Result<bool, SourceError> {
        self.validate_config()?;
        self.http.validate_checkpoint(checkpoint).await
    }
}

#[cfg(test)]
mod delivery_tests;
#[cfg(test)]
#[path = "ws_tests.rs"]
mod tests;
